packages/techdocs-cli/src/commands/generate/generate.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { resolve } from "path";
+import { Command } from "commander";
+import fs from "fs-extra";
+import Docker from "dockerode";
+import {
+  TechdocsGenerator,
+  ParsedLocationAnnotation
+} from "@backstage/techdocs-common";
+import { ConfigReader } from "@backstage/config";
+import {
+  convertTechDocsRefToLocationAnnotation,
+  createLogger
+} from "../../lib/utility";
+
+export default async function generate(cmd: Command) {
+  // Use techdocs-common package to generate docs. Keep consistency between Backstage and CI generating docs.
+  // Docs can be prepared using actions/checkout or git clone, or similar paradigms on CI. The TechDocs CI workflow
+  // will run on the CI pipeline containing the documentation files.
+  //
+  // Create a generator from config
+  // Generator.run (source directory, docker client, parsed annotation, output dir)
+  // Will create output dir parents on its own
+
+  const logger = createLogger({ verbose: cmd.verbose });
+
+  const sourceDir = resolve(cmd.sourceDir);
+  const outputDir = resolve(cmd.outputDir);
+  logger.info(`Using source dir ${sourceDir}`);
+  logger.info(`Will output generated files in ${outputDir}`);
+
+  logger.verbose("Creating output directory if it does not exist.");
+
+  await fs.ensureDir(outputDir);
+
+  const config = new ConfigReader({
+    techdocs: {
+      generators: {
+        techdocs: cmd.docker ? "docker" : "local"
+      }
+    }
+  });
+
+  // Docker client (conditionally) used by the generators, based on techdocs.generators config.
+  const dockerClient = new Docker();
+
+  let parsedLocationAnnotation = {} as ParsedLocationAnnotation;
+  if (cmd.techdocsRef) {
+    try {
+      parsedLocationAnnotation = convertTechDocsRefToLocationAnnotation(
+        cmd.techdocsRef
+      );
+    } catch (err) {
+      logger.error(err.message);
+    }
+  }
+
+  // Generate docs using @backstage/techdocs-common
+  const techdocsGenerator = new TechdocsGenerator(logger, config);
+  logger.info("Generating documentation...");
+  await techdocsGenerator.run({
+    directory: sourceDir,
+    dockerClient: dockerClient,
+    expectedResultDir: outputDir,
+    ...(cmd.techdocsRef && {
```

Comment:
Does this mean the repo is fetched from the ref, or is it going to be possible to generate from a local relative path (eg. `.`)?

If this is to run in CI, it makes most sense to me that it could just build the docs site from the local path since the code is already there. This would allow some custom pre generation prep. If the underlying generate is just pulling the code again this is not possible.

Reply:
Actually, after further look it appears my assumption might have been incorrect on the purpose of this flag 😥. I guess I'm unsure what it actually is used for in the context of the CLI then?

Reply:
You got it right. The `--directory` option is used to build the docs using a local path.

`--techdocs-ref` is optional and will add few extra features in the generated site. As of now, it serves only one purpose, enable "Edit page" icon on pages by adding `repo_url` to `mkdocs.yml`.
https://github.com/backstage/backstage/blob/68ad5af519ff3bbfb5b154309f39ae79e737958c/packages/techdocs-common/src/stages/generate/techdocs.ts#L91-L98

[ ] TODO: Describe more about the purpose of `--techdocs-ref` here.

Reply:
Thanks for clarifying! I'm assuming the generate process ignores any pre-existing `repo_url`, etc. in the `mkdocs.yml` file? Hoping to do a pre run script that generates our `mkdocs.yml` automatically before running `techdocs-cli generate` in CI since we don't store the mkdocs file in source control atm.

Reply:
Yes! If `repo_url` exists in `mkdocs.yml` then it will be left untouched.

I am curious about the `mkdocs.yml` generation on fly and I think it would be a great feature in `techdocs-cli`. Making `mkdocs.yml` a non-requirement is in the roadmap, I think.
Maybe `techdocs-cli generate` can do this too.

Reply:
Yeah, we generate the file from a template in CI. I suppose the generate command could accept a flag to a template URI which then injects variables. The repo_url can be figured out through the got remote in most cases.

Reply:
yes internally mkdocs.y(a)ml is optional for us too so agree with taking away that extra step of config!